# Exercises XP

**Exercise 1: Duplicate Detection And Removal**

Instructions
Objective: Identify and remove duplicate entries in the Titanic dataset.

Load the Titanic dataset.
Identify if there are any duplicate rows based on all columns.
Remove any duplicate rows found in the dataset.
Verify the removal of duplicates by checking the number of rows before and after the duplicate removal.
Hint: Use the duplicated() and drop_duplicates() functions in Pandas.

In [3]:
import pandas as pd

titanic = pd.read_csv('train.csv')

titanic.head()

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   

   Parch            Ticket     Fare Cabin Embarked  
0      0         A/5 21171   7.2500   NaN        S  
1      0          PC 17599  71.2833   C85        C  
2      0  STON/O2. 3101282   7.9250   NaN        S  
3      0            113803  53.1000  C123        S  
4      0            373450   8.0500   NaN        S

In [7]:
print(len(titanic))
print(titanic.duplicated().sum())

891
0


<div style="border:solid green 2px; padding: 20px">
    
 There are no duplicates in this dataset

**Exercise 2: Handling Missing Values**

Instructions
Identify columns in the Titanic dataset with missing values.
Explore different strategies for handling missing data, such as removal, imputation, and filling with a constant value.
Apply each strategy to different columns based on the nature of the data.

In [8]:
missing_data = titanic.isnull()
print(missing_data.head())

missing_counts = titanic.isnull().sum()
print(missing_counts)

   PassengerId  Survived  Pclass   Name    Sex    Age  SibSp  Parch  Ticket  \
0        False     False   False  False  False  False  False  False   False   
1        False     False   False  False  False  False  False  False   False   
2        False     False   False  False  False  False  False  False   False   
3        False     False   False  False  False  False  False  False   False   
4        False     False   False  False  False  False  False  False   False   

    Fare  Cabin  Embarked  
0  False   True     False  
1  False  False     False  
2  False   True     False  
3  False  False     False  
4  False   True     False  
PassengerId      0
Survived         0
Pclass           0
Name             0
Sex              0
Age            177
SibSp            0
Parch            0
Ticket           0
Fare             0
Cabin          687
Embarked         2
dtype: int64


<div style="border:solid green 2px; padding: 20px">
    
There are 687 missing values out of 891 rows in the Cabin column, so this column is uninformative and can be deleted completely

In [9]:
titanic_clean = titanic.drop('Cabin', axis=1)
titanic_clean.head()

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   

   Parch            Ticket     Fare Embarked  
0      0         A/5 21171   7.2500        S  
1      0          PC 17599  71.2833        C  
2      0  STON/O2. 3101282   7.9250        S  
3      0            113803  53.1000        S  
4      0            373450   8.0500        S

<div style="border:solid green 2px; padding: 20px">
    There are 177 missing values in the age column. They can be filled with the median age value.

In [12]:
titanic_clean['Age'] = titanic_clean['Age'].fillna(titanic_clean['Age'].median())
titanic_clean['Age'].isnull().sum()

0

In [13]:
titanic_clean['Embarked'].value_counts()

S    644
C    168
Q     77
Name: Embarked, dtype: int64

<div style="border:solid green 2px; padding: 20px">
    Most of the values in the Embarked column are "S". There are only two missing values in this column, so they can be replaced with "S".

In [14]:
titanic_clean['Embarked'] = titanic_clean['Embarked'].fillna('S')
titanic_clean['Embarked'].isnull().sum()

0

**Exercise 3: Feature Engineering**

Instructions
Create new features, such as ‘Family Size’ from ‘SibSp’ and ‘Parch’, and ‘Title’ extracted from the ‘Name’ column.
Convert categorical variables into numerical form using techniques like one-hot encoding or label encoding.
Normalize or standardize numerical features if required.

In [15]:
# Create new features - ‘Family Size’ from ‘SibSp’ and ‘Parch’
titanic_clean['Family Size'] = titanic_clean['SibSp'] + titanic_clean['Parch']
titanic_clean.head()

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   

   Parch            Ticket     Fare Embarked  Family Size  
0      0         A/5 21171   7.2500        S            1  
1      0          PC 17599  71.2833        C            1  
2      0  STON/O2. 3101282   7.9250        S            0  
3      0            113803  53.1000        S            1  
4      0            373450   8.0500        S            0

In [17]:
# Create new features -‘Title’ extracted from the ‘Name’ column
titanic_clean['Title'] = titanic_clean['Name'].apply(lambda name: name.split(',')[0].split('.')[0].strip())
titanic_clean.head()

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   

   Parch            Ticket     Fare Embarked  Family Size      Title  
0      0         A/5 21171   7.2500        S            1     Braund  
1      0          PC 17599  71.2833        C            1    Cumings  
2      0  STON/O2. 3101282   7.9250        S            0  Heikkinen  
3      0            113803  53.1000        S            1   Futrelle  
4      0            373450   8.0500        S            0      Allen

**Exercise 4: Outlier Detection And Handling**

Instructions
Use statistical methods to detect outliers in columns like ‘Fare’ and ‘Age’.
Decide on a strategy to handle the identified outliers, such as capping, transformation, or removal.
Implement the chosen strategy and assess its impact on the dataset.

In [25]:
Q1 = titanic_clean['Fare'].quantile(0.25)
Q3 = titanic_clean['Fare'].quantile(0.75)
IQR = Q3 - Q1
lower_bound = Q1 - 1.5 * IQR
upper_bound = Q3 + 1.5 * IQR
titanic_clean = titanic_clean[(titanic_clean['Fare'] >= lower_bound) & (titanic_clean['Fare'] <= upper_bound)]

In [28]:
import numpy as np
from scipy.stats import zscore

z_scores_age = np.abs(zscore(titanic_clean['Age']))
outliers_age = titanic_clean[np.abs(z_scores_age) > 3]
print(len(outliers_age))

titanic_clean['Age'] = np.where(np.abs(z_scores_age) > 3, titanic_clean['Age'].median(), titanic_clean['Age'])

6


**Exercise 5: Data Standardization And Normalization**

Instructions
Assess the scale and distribution of numerical columns in the dataset.
Apply standardization to features with a wide range of values.
Normalize data that requires a bounded range, like [0, 1].

In [30]:
# Normalize numerical features
numerical_columns = ['Age', 'Fare', 'Family Size']
scaler = StandardScaler()
titanic_clean[numerical_columns] = scaler.fit_transform(titanic_clean[numerical_columns])
titanic_clean.head()

PassengerId  Survived  Pclass  \
0            1         0       3   
2            3         1       3   
3            4         1       1   
4            5         0       3   
5            6         0       3   

                                           Name  Sex       Age  SibSp  Parch  \
0                       Braund, Mr. Owen Harris    1 -0.526311      1      0   
2                        Heikkinen, Miss. Laina    0 -0.197482      0      0   
3  Futrelle, Mrs. Jacques Heath (Lily May Peel)    0  0.542381      1      0   
4                      Allen, Mr. William Henry    1  0.542381      0      0   
5                              Moran, Mr. James    1 -0.033068      0      0   

             Ticket      Fare  Family Size      Title  Embarked_Q  Embarked_S  
0         A/5 21171 -0.779117     0.156739     Braund           0           1  
2  STON/O2. 3101282 -0.729373    -0.549497  Heikkinen           0           1  
3            113803  2.599828     0.156739   Futrelle           0           1  
4            373450 -0.720161    -0.549497      Allen           0           1  
5            330877 -0.690071    -0.549497      Moran           1           0

**Exercise 6: Feature Encoding**

Instructions
Identify categorical columns in the Titanic dataset, such as ‘Sex’ and ‘Embarked’.
Use one-hot encoding for nominal variables and label encoding for ordinal variables.
Integrate the encoded features back into the main dataset.

In [33]:
# Convert categorical variables into numerical form using one-hot encoding
titanic_clean = pd.get_dummies(titanic_clean, columns=['Embarked'], drop_first=True)
titanic_clean.head()

PassengerId  Survived  Pclass  \
0            1         0       3   
2            3         1       3   
3            4         1       1   
4            5         0       3   
5            6         0       3   

                                           Name  Sex       Age  SibSp  Parch  \
0                       Braund, Mr. Owen Harris    1 -0.526311      1      0   
2                        Heikkinen, Miss. Laina    0 -0.197482      0      0   
3  Futrelle, Mrs. Jacques Heath (Lily May Peel)    0  0.542381      1      0   
4                      Allen, Mr. William Henry    1  0.542381      0      0   
5                              Moran, Mr. James    1 -0.033068      0      0   

             Ticket      Fare  Family Size      Title  Embarked_Q  Embarked_S  
0         A/5 21171 -0.779117     0.156739     Braund           0           1  
2  STON/O2. 3101282 -0.729373    -0.549497  Heikkinen           0           1  
3            113803  2.599828     0.156739   Futrelle           0           1  
4            373450 -0.720161    -0.549497      Allen           0           1  
5            330877 -0.690071    -0.549497      Moran           1           0

In [32]:
# Convert categorical variables into numerical form using label encoding
from sklearn.preprocessing import OneHotEncoder, LabelEncoder, StandardScaler, MinMaxScaler

label_encoder = LabelEncoder()
titanic_clean['Sex'] = label_encoder.fit_transform(titanic_clean['Sex'])

PassengerId  Survived  Pclass  \
0            1         0       3   
2            3         1       3   
3            4         1       1   
4            5         0       3   
5            6         0       3   

                                           Name  Sex       Age  SibSp  Parch  \
0                       Braund, Mr. Owen Harris    1 -0.526311      1      0   
2                        Heikkinen, Miss. Laina    0 -0.197482      0      0   
3  Futrelle, Mrs. Jacques Heath (Lily May Peel)    0  0.542381      1      0   
4                      Allen, Mr. William Henry    1  0.542381      0      0   
5                              Moran, Mr. James    1 -0.033068      0      0   

             Ticket      Fare  Family Size      Title  Embarked_Q  Embarked_S  
0         A/5 21171 -0.779117     0.156739     Braund           0           1  
2  STON/O2. 3101282 -0.729373    -0.549497  Heikkinen           0           1  
3            113803  2.599828     0.156739   Futrelle           0           1  
4            373450 -0.720161    -0.549497      Allen           0           1  
5            330877 -0.690071    -0.549497      Moran           1           0

**Exercise 7: Data Transformation For Age Feature**

Instructions
Create age groups (bins) from the ‘Age’ column to categorize passengers into different age categories.
Apply one-hot encoding to the age groups to convert them into binary features.

In [35]:
bins = [0, 18, 30, 50, 100] 
labels = ['0-17', '18-29', '30-49', '50+']
titanic['AgeGroup'] = pd.cut(titanic['Age'], bins=bins, labels=labels, right=False)

titanic_encoded = pd.get_dummies(titanic, columns=['AgeGroup'], prefix='AgeGroup', drop_first=True)
titanic_encoded.head()

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   

   Parch            Ticket     Fare Cabin Embarked  AgeGroup_18-29  \
0      0         A/5 21171   7.2500   NaN        S               1   
1      0          PC 17599  71.2833   C85        C               0   
2      0  STON/O2. 3101282   7.9250   NaN        S               1   
3      0            113803  53.1000  C123        S               0   
4      0            373450   8.0500   NaN        S               0   

   AgeGroup_30-49  AgeGroup_50+  
0               0             0  
1               1             0  
2               0             0  
3               1             0  
4               1             0